### PR TITLE
implemented count function

### PR DESCRIPTION
**Pull request by Aman-Godara**
Counts the number of time substring `pattern` has occurred in the input `string`.

Tasks

- [x] implemented count function
- [x] added test cases
- [x] documented the function
- [x] Elemental or Pure -> Elemental

### CODE REVIEW

File: src/stdlib_strings.f90 [hunk 1]
```diff
@@ -649,4 +660,87 @@ pure function replace_all_char_char_char(string, pattern, replacement) result(re
...
+        if (length_pattern > 0 .and. length_pattern <= length_string) then
+            lps_array = compute_lps(pattern)
+
+            s_i = 1
+            p_i = 1
+            do while(s_i <= length_string)
+                if (string(s_i:s_i) == pattern(p_i:p_i)) then
+                    if (p_i == length_pattern) then
+                        res = res + 1
+                        if (consider_overlapping_) then
```
Comment (milancurcic) [inline src/stdlib_strings.f90:727]: `consider_overlapping_` is used only once. So we may as well take advantage of `optval()` to full effect and drop the temporary variable `consider_overlapping_`:

```suggestion

        res = 0
        length_string = len(string)
        length_pattern = len(pattern)

        if (length_pattern > 0 .and. length_pattern <= length_string) then
            lps_array = compute_lps(pattern)

            s_i = 1
            p_i = 1
            do while(s_i <= length_string)
                if (string(s_i:s_i) == pattern(p_i:p_i)) then
                    if (p_i == length_pattern) then
                        res = res + 1
                        if (optval(consider_overlapping, .true.)) then
```

File: src/stdlib_strings.f90 [hunk 2]
```diff
@@ -649,4 +660,87 @@ pure function replace_all_char_char_char(string, pattern, replacement) result(re
...
+        res = 0
+        length_string = len(string)
+        length_pattern = len(pattern)
+
+        if (length_pattern > 0 .and. length_pattern <= length_string) then
+            lps_array = compute_lps(pattern)
+
+            s_i = 1
+            p_i = 1
+            do while(s_i <= length_string)
```
Comment (milancurcic) [inline src/stdlib_strings.f90:723]: For consistency with other constructs

```suggestion
            do while (s_i <= length_string)
```

**Review by milancurcic — APPROVED** [review]
Looks good, thank you!

File: src/stdlib_strings.f90 (hunk 1 above)
Comment (Aman-Godara) [reply src/stdlib_strings.f90]: done!

File: doc/specs/stdlib_strings.md [hunk 3]
```diff
@@ -381,4 +381,56 @@ program demo_replace_all
...
\ No newline at end of file
+```
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### `count`
+
+#### Description
+
+Returns the number of times substring `pattern` has occurred in the input string `string`.
```
Comment (jvdp1) [inline doc/specs/stdlib_strings.md:392]:
```suggestion
Returns the number of times the substring `pattern` has occurred in the input string `string`.  
```

**Review by awvwgk — APPROVED** [review]
Thanks, looks good to me.

File: src/stdlib_strings.f90 [hunk 4]
```diff
@@ -93,6 +93,17 @@ module stdlib_strings
         module procedure :: replace_all_char_char_char
     end interface replace_all
 
+    !> Returns the number of times substring 'pattern' has appeared in the
+    !> input string 'string'
+    !> [Specifications](../page/specs/stdlib_strings.html#count)
+    !> Version: experimental
```
Comment (jvdp1) [inline src/stdlib_strings.f90:99]: Version can be considered as meta-data by ford (https://github.com/Fortran-FOSS-Programmers/ford/wiki/Documentation-Meta-Data)
```suggestion
    !> Version: experimental
    !>
    !> Returns the number of times substring 'pattern' has appeared in the
    !> input string 'string'
    !> [Specifications](../page/specs/stdlib_strings.html#count)
```

**Review by jvdp1 — APPROVED** [review]
Pending a few minor comments, SGTM.
Thank you.

**Pull request merged**